vcd/resource_vcd_data_center_group.go

```diff
@@ -0,0 +1,452 @@
+	}
+
+	vdcGroupConfig := getDataCenterGroupConfigurationType(d)
+	createdVdcGroup, err := adminOrg.CreateNsxtVdcGroup(vdcGroupConfig.Name, vdcGroupConfig.Description, vdcGroupConfig.StartingVdcId, vdcGroupConfig.ParticipatingVdcIds)
+	if err != nil {
+		return diag.Errorf("error creating data center group: %s", err)
+	}
+
+	if d.Get("dfw_enabled").(bool) {
+		createdVdcGroup, err = createdVdcGroup.ActivateDfw()
+		if err != nil {
+			return diag.Errorf("error enabling DFW for data center group: %s", err)
+		}
+		// by default, default policy will be enabled when DFW is enabled, so only need code for disabling
+		if !d.Get("default_policy_status").(bool) {
+			createdVdcGroup, err = createdVdcGroup.DisableDefaultPolicy()
+			if err != nil {
+				return diag.Errorf("error disabling default policy for data center group: %s", err)
+			}
+		}
+	}
+
+	d.SetId(createdVdcGroup.VdcGroup.Id)
+	return resourceVcdDataCenterGroupRead(ctx, d, meta)
+}
+
+// resourceVcdDataCenterGroupUpdate covers Update functionality for resource
+func resourceVcdDataCenterGroupUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	adminOrg, err := vcdClient.GetAdminOrgFromResource(d)
+	if err != nil {
+		return diag.Errorf(errorRetrievingOrg, err)
+	}
+
+	vdcGroup, err := adminOrg.GetVdcGroupById(d.Id())
+	if err != nil {
+		return diag.Errorf("[data center group update] : %s", err)
+	}
+
+	if d.HasChanges("name", "description", "participating_vdc_ids") {
+		vdcGroupConfig := getDataCenterGroupConfigurationType(d)
+		vdcGroup, err = vdcGroup.Update(vdcGroupConfig.Name, vdcGroupConfig.Description, vdcGroupConfig.ParticipatingVdcIds)
+		if err != nil {
+			return diag.Errorf("[data center group update] : %s", err)
+		}
+	}
+
+	if d.HasChange("dfw_enabled") {
+		if d.Get("dfw_enabled").(bool) {
+			vdcGroup, err = vdcGroup.ActivateDfw()
+		} else {
+			vdcGroup, err = vdcGroup.DisableDefaultPolicy()
+			// ignore if it isn't possible to change
+			if err != nil && err.Error() != "DFW has to be enabled before changing Default policy" {
+				return diag.Errorf("error disabling default policy for data center group: %s", err)
+			}
+			vdcGroup, err = vdcGroup.DeActivateDfw()
+		}
+		if err != nil {
+			return diag.Errorf("error activating/deactivating DFW for data center group: %s", err)
+		}
+	}
+
+	if d.HasChange("default_policy_status") {
+		errDiag := applyDefaultPolicy(d, vdcGroup)
+		if errDiag != nil {
+			return errDiag
+		}
+	}
+
+	return resourceVcdDataCenterGroupRead(ctx, d, meta)
+}
+
+func applyDefaultPolicy(d *schema.ResourceData, vdcGroup *govcd.VdcGroup) diag.Diagnostics {
+	var err error
+	if !d.Get("default_policy_status").(bool) {
+		_, err = vdcGroup.DisableDefaultPolicy()
+	} else {
+		_, err = vdcGroup.EnableDefaultPolicy()
+	}
+	// ignore if it isn't possible to change
+	if err != nil && err.Error() != "DFW has to be enabled before changing Default policy" {
```

Comment:
This line will mask error `DFW has to be enabled before changing Default policy` from user forever because we return `nil` below this line. Is it the goal?

Reply:
Yes, it is the intention. You can't change value DFW if it is disabled, so in terraform perspective you change the value but maybe there isn't a possibility to make a change in API.